File: src/main/java/ru/vood/freemarker/ext/sql/model/FetchedResultSetRowModel.java
package ru.vood.freemarker.ext.sql.model;

import freemarker.template.*;
import ru.vood.freemarker.ext.sql.FetchedResultSet;

public class FetchedResultSetRowModel extends WrappingTemplateModel implements TemplateSequenceModel, TemplateHashModelEx {
    private final FetchedResultSet resultSet;
    private final int rowIndex;

    FetchedResultSetRowModel(FetchedResultSet frs, int index, ObjectWrapper wrapper) {
        super(wrapper);
        this.resultSet = frs;
        this.rowIndex = index;
    }

    public TemplateModel get(int index) throws TemplateModelException {
        return wrap(resultSet.getRows().get(rowIndex).get(index));
    }

    public TemplateModel get(String key) throws TemplateModelException {
        return wrap(resultSet.getRows().get(rowIndex).get(resultSet.getColumnIndex(key)));
    }

    public int size() {
        return resultSet.getColumnLabels().size();
    }

    public TemplateCollectionModel keys() {
        return new SimpleCollection(resultSet.getColumnLabels(), getObjectWrapper());
    }

    public TemplateCollectionModel values() {
        return new SimpleCollection(resultSet.getRows().get(rowIndex), getObjectWrapper());
    }

    public boolean isEmpty() {
        return false;
    }
}

File: src/main/java/ru/vood/freemarker/ext/sql/model/FetchedResultSetRowModel.kt
package ru.vood.freemarker.ext.sql.model

import freemarker.template.*
import ru.vood.freemarker.ext.sql.FetchedResultSet

/**
 * This class wraps [java.sql.ResultSet]'s row and adapts it for using in FTL both as a sequence and a hash of
 * rows.
 */
class FetchedResultSetRowModel internal constructor(
        private val resultSet: FetchedResultSet,
        private val rowIndex: Int, wrapper: ObjectWrapper) : WrappingTemplateModel(wrapper), TemplateSequenceModel, TemplateHashModelEx {

    override fun get(index: Int): TemplateModel {
        return wrap(resultSet.rows[rowIndex][index])
    }

    override fun get(key: String): TemplateModel {
        return wrap(resultSet.rows[rowIndex][resultSet.getColumnIndex(key)])
    }

    override fun size(): Int {
        return resultSet.columnLabels.size
    }

    override fun keys(): TemplateCollectionModel {
        return SimpleCollection(resultSet.columnLabels, objectWrapper)
    }

    override fun values(): TemplateCollectionModel {
        return SimpleCollection(resultSet.rows[rowIndex], objectWrapper)
    }

    override fun isEmpty(): Boolean {
        return false
    }

}